### ISSUE TITLE
TS2678 Enum "not comparable" but equal

### ISSUE THREAD

**Issue by pfusik** [microsoft/TypeScript#52407]
# Bug Report

### 🔎 Search Terms

- TS2678
- is not comparable to type

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about TS2678

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.9.4#code/KYOwrgtgBAKg9ga1FA3gWAFBSgUQPIBiANJtgJIAiJWUZAcgMIBKOAsjnTNdhTs2xxiYAvpkwBjOCADOAFyizEoaVAC8UANqwlIAHSUi2pHvr92nKAF0A3JgA2weeLAAnF6FnxjUAFxHQthiYAGZgIOKyAJZSUA4AHgAUAJSY6DTObh5eyOqKxtK60gAWkcGyyVAA-JX+evgEgaJBGKHhUTEADgCGLtLAOHEdLsmppFDSAO6RsuJFUAkZ7iCeOkmoY+JdfbX6jCzmMD5j2JIycA66dnAA5gkAREPAkeHuEB53SYHY2PHJX9-dXr9QbDT7HKDuWSuED-AAmwGCXTAdlkRxoJyk0nOwEuN3ucFkRWALg+-x+wESYPRUAARu4ugh-k1sME4C55tZrGs0t9xlMZnMFq4litjNzwZtttk6oQ0bzsJDoWSoJLgDtTPtBHL5acsRcrrcHnA5M9xK93lT5bEKX9wdg6cAGcrVTteGYtXaVZjsbjDR1jbJ4WbgG9lqTPb9LfKHU7wfDEcjUZ7dT6DfccEwmHgmB9PYqXDDwcyoE0mvYbVTAX0BkNbRggA)

### 💻 Code

```ts
enum Token {
  EOF,
  ID,
  INCREMENT,
  DECREMENT
}

const tokens = [ Token.ID, Token.INCREMENT ];
let currentToken : Token;

function lex()
{
  currentToken = tokens.shift() ?? Token.EOF;
}

function parseExpr()
{
  switch (currentToken) {
  case Token.INCREMENT:
    console.log("preincrement");
    lex();
    parseExpr();
    return;
  default:
    console.log("other");
    lex();
    break;
  }
  for (;;) {
    switch (currentToken) {
    case Token.EOF:
      return;
    case Token.INCREMENT:
      console.log("postincrement");
      lex();
      break;
    case Token.DECREMENT:
      console.log("postdecrement");
      lex();
      break;
    default:
      console.log("ERROR")
      return;
    }
  }
}

lex();
parseExpr();
```

### 🙁 Actual behavior

    loop-switch-enum.ts:33:10 - error TS2678: Type 'Token.INCREMENT' is not comparable to type 'Token.EOF | Token.ID | Token.DECREMENT'.

    33     case Token.INCREMENT:
                ~~~~~~~~~~~~~~~

js is emitted and prints:

    other
    postincrement

so the `case` is matched.

### 🙂 Expected behavior

Code compiles successfully.

**Comment by MartinJohns** [comment]
Essentially a duplicate of #9998. In your code you check for `Token.INCREMENT` at line 19 and return in this case in line 23. From this the compiler gathers that later in the method the type can **not** be `Token.INCREMENT` anymore. It does not see that `lex()` could change the value of `currentToken` due to reasons described in #9998.

**Comment by pfusik** [comment]
That's a long thread. What is the workaround?

**Comment by jcalz** [comment]
One workaround is to re-widen `currentToken` to `Token`  when you `switch`, like `switch (currentToken as Token)` as shown [here](https://www.typescriptlang.org/play?ts=4.9.4#code/KYOwrgtgBAKg9ga1FA3gWAFBSgUQPIBiANJtgJIAiJWUZAcgMIBKOAsjnTNdhTs2xxiYAvpkwBjOCADOAFyizEoaVAC8UANqwlIAHSUi2pHvr92nKAF0A3JgA2weeLAAnF6FnxjUAFxHQthiYAGZgIOKyAJZSUA4AHgAUAJSY6DTObh5eyOqKxtK60gAWkcGyyVAA-JX+evgEgaJBGKHhUTEADgCGLtLAOHEdLsmppFDSAO6RsuJFUAkZ7iCeOkmoY+JdfbX6jCzmMD5j2JIycA66dnAA5gkAREPAkeHuEB53SYHY2PHJX9-dXr9QbDT7HKDuWSuED-AAmwGCXTAdlkRxoJyk0nOwEuN3ucFkRWALg+-x+wESYPRUAARu4ugh-k1sME4C55tZrGs0t9xlMZnMFq4litvFtatzwZtttk6oQ0bzsJDoWSoNLgDtTPtBArFacsRcrrcHnA5M9xK93lTFbEKX9wdg6cAGar1TteGYdQ61ZjsbjjR1TbJ4RbgG9lqTvb9rYqnS7wfDEcjUd79X6jfccEwmHgmB9vcqXDDwcyoE0mvY7VTAX0BkN7RggA).

**Comment by fatcerberus** [comment]
> That's a long thread.

tl;dr: Control flow-based type narrowing is:
* Optimistic about function/method calls.  Most calls won't mutate local state and losing all local narrowings every time *any* function is called would render the feature basically worthless.
* Pessimistic about closures.  The compiler doesn't know when a callback will actually run.

Absent the compiler actually stepping into function calls to see what happens inside, this is the best TS can do today.

**Comment by RyanCavanaugh** [comment]
Incredibly good TL;DR actually, maybe I want to put that at the top of 9998
